### ISSUE TITLE
为什么在算assumed t+1时的θ时要自己去实现优化器的更新操作

### ISSUE THREAD

**Issue by mztkenan** [LaceyChen17/lambda-opt#3]
你好，我把源代码看了一遍，我不太懂为什么在算assumed t+1时的θ时，要搞这么麻烦，还要自己去实现优化器的更新操作。在part 2中直接加入regularization ，得到的更新参数就是assumed t+1时的θ，为什么要分开来，自己手动更新。

**Comment by yihong-chen** [comment]
This is to make `optimizer.step` differentiable with respect to $\lambda$. In order to make sure the gradient path of $\lambda$ update doesn't corrupt that of $\theta$ update, we implemented them separately.  

I admit that the optimizer update by hand is quite annoying. One way to bypass this trouble would be using [Higher](https://github.com/facebookresearch/higher), which makes pytorch optimizer differentiable. If you are interested in this, we can work on it together and get rid of this troublesome manual optimizer update. :smiley:

**Comment by mztkenan** [comment]
感谢回复，您的意思是说因为对于λ来说因为有clamp 非负的限制所以是不可微的，所以把loss_bpr和loss_reg分开来计算，不然会影响loss_bpr这边梯度的计算。

**Comment by mztkenan** [comment]
但是实际上，在得到的θ(t+1)时，此时λ认为是定值常数，只更新θ，我认为这里没有影响，不知道哪里想的不对。例如在代码SGD里，手工设next_emb_non_reg - 2 * self.mf_lr * l * curr_emb，此时l是定值。和优化器自动更新是没有区别的。

**Comment by yihong-chen** [comment]
Thanks for your continued interest.

> 感谢回复，您的意思是说因为对于λ来说因为有clamp 非负的限制所以是不可微的，所以把loss_bpr和loss_reg分开来计算，不然会影响loss_bpr这边梯度的计算。

It is not exactly due to `clamp` and the non-negative constraints of $\lambda$.

>  但是实际上，在得到的θ(t+1)时，此时λ认为是定值常数，只更新θ，我认为这里没有影响，不知道哪里想的不对。例如在代码SGD里，手工设next_emb_non_reg - 2 * self.mf_lr * l * curr_emb，此时l是定值。和优化器自动更新是没有区别的。

The $l$ here is differentiable, which is quite different from the usual optimizer update. The key for $\lambda$-update is to make the process of obtaining $\Theta_{t+1}$ differentiable with respect to $\lambda$. If you check the above $l$, you will find that it is actually learnable.

I hope this answer would be helpful to you.

**Comment by mztkenan** [comment]
感谢回复，那可能我这里理解有点问题，在use_optimizer函数里设置了 optimizer = torch.optim.SGD(network.parameters())，我以为pytorch这里设置了optimizer的优化变量，l在这里就变为常数不参与梯度计算了。在TensorFlow里是可以让λ为常数不参与梯度计算的。

**Comment by mztkenan** [comment]
我理解的得到λ(t+1)的两步中，第一步训练集上更新获得θ(t+1)，此时认为λ是常数。第二步，认为θ(t+1)是常数，在验证集上更新λ得到过λ(t+1)。可能我哪里理解的有误，不知道您为何要在第一步获得θ(t+1)的时候让λ是可微的。

**Comment by yihong-chen** [comment]
> 我理解的得到λ(t+1)的两步中，第一步训练集上更新获得θ(t+1)，此时认为λ是常数。第二步，认为θ(t+1)是常数，在验证集上更新λ得到过λ(t+1)。可能我哪里理解的有误，不知道您为何要在第一步获得θ(t+1)的时候让λ是可微的。

You are quite right. $\lambda$ is only differentiable for the second step. assumed $\theta_{t+1}$ is only used for the second step. The process of obtaining assumed $\theta_{t+1}$ should be differentiable with respect to $\lambda$. That's why we don't use the $\theta_{t+1}$ in the 1st step to update $\lambda$.

**Comment by mztkenan** [comment]
感谢作者回复，在使用TensorFlow复现的过程中我明白了，θ(t+1)不是一个固定的值，而是一个可以关于λ可以求导的op。所以要把更新这里分开来。

**Comment by mztkenan** [comment]
另外有个问题就是由于文章采用的是矩阵分解，由θ得到loss值的过程只要item_emb和user_emb相乘即可。但是应用到神经网络中的话，如果θ是op不是variable的话不知道如何操作得到神经网络的loss。不知道作者有没有想过扩展这个框架能适用于通用情况。

**Comment by yihong-chen** [comment]
It looks interesting. Could you give an example of "op instead of variable"?

**Comment by mztkenan** [comment]
例如doc embedding 的θ(t+1),不能是已经更新完的Variable，而是op计算出来的Tensor，这样才能使得λ是learnable来更新λ。

已解决，写了个装饰器来包裹tf.get_variable，根据variable_flag返回Tensor或者Variable，在得到θ(t+1)的时候得到op计算出来的Tensor。这样无需更改原有使用tf.get_variable的代码。

Demo如下
```
class Dlp_Rank(object):
    def __init__(self):
        self.variable_flag = True
        self.theta_tensor_dict={}

    def decorator_variable(self,f):
        zbb=tf.constant(1)
        theta_tensor_dict={"zbb-copy":zbb} # TO FIX
        def get_variable(name,*args, **kwargs):
            if self.variable_flag: return f(name, *args, **kwargs)
            else:return theta_tensor_dict[name+"-copy"]
        tf.logging.info("Decorate at first")
        return get_variable

dlp=Dlp_Rank()
tf.get_variable=dlp.decorator_variable(tf.get_variable)
dlp.variable_flag=False
ccc=tf.get_variable("zbb",dtype=tf.int32,initializer=2)
# 此时ccc是Tensor
```

**Comment by yihong-chen** [comment]
Good job! Thank you for the sharing.
